Level Of Detail

KatsuroDev opened the pull request:
Level of detail: 0
![image](https://github.com/user-attachments/assets/64f7c63e-15f9-48d9-b8a5-d0e4f14110fc)

Level of detail: 1
![image](https://github.com/user-attachments/assets/5e6f4b9b-d008-4892-8ed3-4de82dc5655e)

KatsuroDev committed:
Add Height Curve

KatsuroDev committed:
Add LOD WIP

Elatronion commented on Scripts/Chunk.gd:
```diff
@@ -28,28 +29,30 @@ func _generate_surface() -> Array:
+	for z in range(0, chunk_data.size, mesh_simplification_increment):
+		for x in range(0, chunk_data.size, mesh_simplification_increment):
+			var index: int = (x / mesh_simplification_increment) + (z / mesh_simplification_increment) * vertices_per_line
 			var height: float = chunk_data.height_map[index]
```
This is incorrect, as you height_map is still full sized.
You need to either update how you access the index of the height_map:
```
var height_index: int = x + z * chunk_data.width
```

Or have the LOD be considered to update the height_map array resolution to match the vertices.

Elatronion commented on Scripts/Chunk.gd:
```diff
@@ -28,28 +29,30 @@ func _generate_surface() -> Array:
 	var indices := PackedInt32Array()
 	var uvs := PackedVector2Array()
 	var triangle_index: int = 0
+	var mesh_simplification_increment: int = 1 if level_of_detail == 0 else level_of_detail * 2
+	var vertices_per_line: int = (chunk_data.size - 1) / mesh_simplification_increment + 1
```
At LODs > 3, the mesh disappears.
This should be fixed.

At worst, if it's impossible to simplify the mesh further, you may still want a minimum of 4 vertices to form a quad.

Elatronion reviewed (changes requested):
You need to work on data alignment, and take a look LODs that cause the terrain to vanish.

KatsuroDev committed:
Fix Height Indexing And Exceeding LOD

KatsuroDev committed:
Clamp LOD Based On Mesh Size

- Calculate the number of factors of the mesh size to determine the max amount of LOD.

Elatronion commented:
With Level of Detail set to 3, upon generation I get the following error.

```
ERROR: res://Scripts/Chunk.gd:54 - Out of bounds set index '24' (on base: 'PackedInt32Array')
```

Every other level works, it's just 3 that crashes.

I could take a deeper look later, perhaps together, but for now it's not obvious to me what the issue is.
My best guess is some division issue. As in: you can't properly split up your vertices.

A good approach could be to start with the lowest resolution LOD and work up from there.
So, rather than trying to simplify the mesh, you can just add complexity through subdivision.

KatsuroDev commented:
Yes we should try and investigate this together.

KatsuroDev commented:
Can you make an issue for this?
